In [1]:
# importing data manipulation libraries
import pandas as pd
import numpy as np
# from ydata_profiling import ProfileReport

# importing visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns

# configure notebook for inline plotting
%matplotlib inline

# config pandas to display more than 20 columns
pd.set_option('display.max_columns',250)

# set grid style 
sns.set_style('darkgrid')

# Data Understanding
The objective in this section is to learn as much as possible about the our match data. Here, we will get a high-level practical understnading of our data and by the end of the process, we should have a clear undertanding of the structure of the dataset, how to clean the data, the target variables and possible modelling techniques.

In [2]:
# read data into pandas dataframe
data = pd.read_csv('./data/Matches.csv', header=0)

In [3]:
# check the shape of dataframe
data.shape

(6460, 215)

The dataset consists of 6,460 observations and 215 features some of which are our target variables.

In [4]:
# check types of columns
data.dtypes.unique()

array([dtype('int64'), dtype('O'), dtype('float64'), dtype('bool')],
      dtype=object)

The result indicates that our dataset consists of categorical (object), boolean and numeric (float and int) features. Before proceeding further, let's confirm that the dtypes match what is expected in the documentation.

### To do (Create table of columns and their definitions) - Use datawrapper and insert link 



## columns to be used in modelling 
- total goal count (will be target variable for objective two)
- team_a_cards_num	team_b_cards_num(objective three)
- odds_ft_1	odds_ft_x	odds_ft_2
- winningTeam (objective one )- re-code column to 1x2
- home_ppg	away_ppg
- team_a_xg_prematch	team_b_xg_prematch	total_xg_prematch

target columns 
- total goal count
- total cards(obtain by summing team a and b cards)
- winningTeam 1x2

features
- odds_ft_1	odds_ft_x	odds_ft_2
- home_ppg	away_ppg
- team_a_xg_prematch	team_b_xg_prematch	total_xg_prematch


In [34]:
# inspect the first few rows
data.head(20)

id  homeID  awayID     season    status  roundID  game_week  \
0   2155     150     108  2016/2017  complete       19          1   
1   2156     145     154  2016/2017  complete       19          1   
2   2157     143     142  2016/2017  complete       19          1   
3   2158     144      92  2016/2017  complete       19          1   
4   2159     147     141  2016/2017  complete       19          1   
5   2160     146     155  2016/2017  complete       19          1   
6   2161      93     156  2016/2017  complete       19          1   
7   2162     148     149  2016/2017  complete       19          1   
8   2163      59     151  2016/2017  complete       19          1   
9   2164     152     153  2016/2017  complete       19          1   
10  2165     149     146  2016/2017  complete       19          2   
11  2166     141      93  2016/2017  complete       19          2   
12  2167     154     150  2016/2017  complete       19          2   
13  2168      92     143  2016/2017  complete       19          2   
14  2169     155     152  2016/2017  complete       19          2   
15  2170     142     144  2016/2017  complete       19          2   
16  2171     145     151  2016/2017  complete       19          2   
17  2172     108      59  2016/2017  complete       19          2   
18  2173     156     147  2016/2017  complete       19          2   
19  2174     153     148  2016/2017  complete       19          2   

    revised_game_week           homeGoals                   awayGoals  \
0                  -1      ['45+1', '57']                      ['47']   
1                  -1                  []                      ['82']   
2                  -1                  []                      ['74']   
3                  -1               ['5']                      ['59']   
4                  -1              ['11']                      ['67']   
5                  -1              ['58']                       ['9']   
6                  -1         ['4', '87']                      ['71']   
7                  -1              ['69']          ['40', '59', '64']   
8                  -1  ['31', '64', '75']  ['45+1', '49', '56', '63']   
9                  -1        ['47', '89']                      ['77']   
10                 -1        ['36', '52']                          []   
11                 -1              ['49']  ['27', '36', '86', '90+5']   
12                 -1                  []              ['79', '90+2']   
13                 -1              ['82']                          []   
14                 -1              ['55']                ['80', '87']   
15                 -1               ['9']              ['45+2', '60']   
16                 -1         ['2', '37']                          []   
17                 -1                  []                          []   
18                 -1              ['71']                ['13', '45']   
19                 -1              ['85']                          []   

    homeGoalCount  awayGoalCount  totalGoalCount  team_a_corners  \
0               2              1               3               5   
1               0              1               1               7   
2               0              1               1               3   
3               1              1               2               5   
4               1              1               2               9   
5               1              1               2               6   
6               2              1               3              10   
7               1              3               4               4   
8               3              4               7               5   
9               2              1               3               7   
10              2              0               2               4   
11              1              4               5               7   
12              0              2               2               8   
13              1              0               1              10   
14

In [ ]:
#columns to be used for modelling 
#goals
#


In [6]:
# inspect the last rows
data.tail()

id  homeID  awayID     season      status  roundID  game_week  \
6455  6689328     143     158  2023/2024  incomplete   100543         38   
6456  6689329     151     223  2023/2024  incomplete   100543         38   
6457  6689330     271     162  2023/2024  incomplete   100543         38   
6458  6689331      93     153  2023/2024  incomplete   100543         38   
6459  6689332     251      92  2023/2024  incomplete   100543         38   

      revised_game_week homeGoals awayGoals  homeGoalCount  awayGoalCount  \
6455                 -1        []        []              0              0   
6456                 -1        []        []              0              0   
6457                 -1        []        []              0              0   
6458                 -1        []        []              0              0   
6459                 -1        []        []              0              0   

      totalGoalCount  team_a_corners  team_b_corners  totalCornerCount  \
6455               0              -1              -1                -1   
6456               0              -1              -1                -1   
6457               0              -1              -1                -1   
6458               0              -1              -1                -1   
6459               0              -1              -1                -1   

      team_a_offsides  team_b_offsides  team_a_yellow_cards  \
6455               -1               -1                   -1   
6456               -1               -1                   -1   
6457               -1               -1                   -1   
6458               -1               -1                   -1   
6459               -1               -1                   -1   

      team_b_yellow_cards  team_a_red_cards  team_b_red_cards  \
6455                   -1                -1                -1   
6456                   -1                -1                -1   
6457                   -1                -1                -1   
6458                   -1                -1                -1   
6459                   -1                -1                -1   

      team_a_shotsOnTarget  team_b_shotsOnTarget  team_a_shotsOffTarget  \
6455                    -1                    -1                     -1   
6456                    -1                    -1                     -1   
6457                    -1                    -1                     -1   
6458                    -1                    -1                     -1   
6459                    -1                    -1                     -1   

      team_b_shotsOffTarget  team_a_shots  team_b_shots  team_a_fouls  \
6455                     -1            -1            -1            -1   
6456                     -1            -1            -1            -1   
6457                     -1            -1            -1            -1   
6458                     -1            -1            -1            -1   
6459                     -1            -1            -1            -1   

      team_b_fouls  team_a_possession  team_b_possession  refereeID  \
6455            -1                 -1                 -1        NaN   
6456            -1                 -1                 -1        NaN   
6457            -1                 -1                 -1        NaN   
6458            -1                 -1                 -1        NaN   
6459            -1                 -1                 -1        NaN   

      coach_a_ID  coach_b_ID                           stadium_name  \
6455        -1.0        -1.0                 Selhurst Park (London)   
6456        -1.0        -1.0                    Anfield (Liverpool)   
6457        -1.0        -1.0  Kenilworth Road (Luton, Bedfordshire)   
6458        -1.0        -1.0            Etihad Stadium (Manchester)   
6459        -1.0        -1.0               Bramall Lane (Sheffield)   

     stadium_location  team_a_cards_num  team_b_cards_num  odds_ft_1  \
6455              NaN                -1                -1   

In [7]:
# inspect random sample of rows
data.sample(5)

id  homeID  awayID     season    status  roundID  game_week  \
4540   754008     149      59  2008/2009  complete    53612          0   
4812   755386     147     162  2007/2008  complete    53621          0   
5139  1056605     143     153  2020/2021  complete    58864         20   
2915   579256     152      92  2019/2020  complete    50055         27   
1803    49250     108     148  2017/2018  complete      243         29   

      revised_game_week      homeGoals          awayGoals  homeGoalCount  \
4540                 -1             []                 []              0   
4812                 -1         ['11']                 []              1   
5139                 -1  ['3', '90+7']  ['9', '25', '65']              2   
2915                 -1   ['15', '48']             ['89']              2   
1803                 -1       ['90+7']             ['35']              1   

      awayGoalCount  totalGoalCount  team_a_corners  team_b_corners  \
4540              0               0              -1              -1   
4812              0               1              -1              -1   
5139              3               5               3               4   
2915              1               3               3               5   
1803              1               2              13               1   

      totalCornerCount  team_a_offsides  team_b_offsides  team_a_yellow_cards  \
4540                -1               -1               -1                    0   
4812                -1               -1               -1                    1   
5139                 7                0                2                    0   
2915                 8                1                0                    1   
1803                14                3                2                    2   

      team_b_yellow_cards  team_a_red_cards  team_b_red_cards  \
4540                    5                 0                 0   
4812                    2                 0                 0   
5139                    0                 0                 0   
2915                    2                 0                 0   
1803                    1                 0                 0   

      team_a_shotsOnTarget  team_b_shotsOnTarget  team_a_shotsOffTarget  \
4540                    -1                    -1                     -1   
4812                    -1                    -1                     -1   
5139                     4                     7                      3   
2915                     8                     4                     10   
1803                     5                     4                      9   

      team_b_shotsOffTarget  team_a_shots  team_b_shots  team_a_fouls  \
4540                     -1            -1            -1            -1   
4812                     -1            -1            -1            -1   
5139                     10             7            17             9   
2915                      2            18             6            -1   
1803                      4            14             8            11   

      team_b_fouls  team_a_possession  team_b_possession  refereeID  \
4540            -1                 -1                 -1        NaN   
4812            -1                 -1                 -1        NaN   
5139             8                 54                 46      705.0   
2915            -1                 49                 51      393.0   
1803             7                 55                 45      726.0   

      coach_a_ID  coach_b_ID                       stadium_name  \
4540     22031.0       145.0          Old Trafford (Manchester)   
4812     10655.0      3610.0  Riverside Stadium (Middlesbrough)   
5139      3610.0       206.0             Selhurst Park (London)   
2915     11725.0       208.0           Stamford Bridge (London)   
1803       204.0       207.0                 King Power Stadium   

                            stadium_location  team_a_cards_num  \
4540              

### Columns of interest:
This were are columns that were identified to contain interesting, problematic or missing data and warranted further inspection.

#### Status Column
The first column of interest is the status colum which indicates the whether a match was completed. Looking at the head, tail and random sample snapshots of the df, the status of a match can be either complete or incomplete. Let's inspect unique values in the column:

In [8]:
# check unique values in the status column
data['status'].unique()

array(['complete', 'incomplete', 'suspended'], dtype=object)

Based on the output above, match status can `complete`, `incomplete`, or `suspended`. The dataframe snapshots, particularly the tail, shows that most of the incomplete matches are from the 2023/2024 season. This makes sense because, the 2023/2024 is ongoing and there are bound to be incomplete matches. For our analysis we will want to drop entries where the match status. 

In [9]:
# Check list of incomplete matches
data[data['status'] == 'incomplete' ].sample(10)

id  homeID  awayID     season      status  roundID  game_week  \
6287  6689160     149      92  2023/2024  incomplete   100543         21   
6363  6689236     143     157  2023/2024  incomplete   100543         29   
6376  6689249     157     153  2023/2024  incomplete   100543         30   
6296  6689169     223     149  2023/2024  incomplete   100543         22   
6319  6689192     223     218  2023/2024  incomplete   100543         24   
6455  6689328     143     158  2023/2024  incomplete   100543         38   
6440  6689313     148     218  2023/2024  incomplete   100543         37   
6372  6689245     218     149  2023/2024  incomplete   100543         30   
6306  6689179     144      92  2023/2024  incomplete   100543         23   
6337  6689210     149     162  2023/2024  incomplete   100543         26   

      revised_game_week homeGoals awayGoals  homeGoalCount  awayGoalCount  \
6287                 -1        []        []              0              0   
6363                 -1        []        []              0              0   
6376                 -1        []        []              0              0   
6296                 -1        []        []              0              0   
6319                 -1        []        []              0              0   
6455                 -1        []        []              0              0   
6440                 -1        []        []              0              0   
6372                 -1        []        []              0              0   
6306                 -1        []        []              0              0   
6337                 -1        []        []              0              0   

      totalGoalCount  team_a_corners  team_b_corners  totalCornerCount  \
6287               0              -1              -1                -1   
6363               0              -1              -1                -1   
6376               0              -1              -1                -1   
6296               0              -1              -1                -1   
6319               0              -1              -1                -1   
6455               0              -1              -1                -1   
6440               0              -1              -1                -1   
6372               0              -1              -1                -1   
6306               0              -1              -1                -1   
6337               0              -1              -1                -1   

      team_a_offsides  team_b_offsides  team_a_yellow_cards  \
6287               -1               -1                   -1   
6363               -1               -1                   -1   
6376               -1               -1                   -1   
6296               -1               -1                   -1   
6319               -1               -1                   -1   
6455               -1               -1                   -1   
6440               -1               -1                   -1   
6372               -1               -1                   -1   
6306               -1               -1                   -1   
6337               -1               -1                   -1   

      team_b_yellow_cards  team_a_red_cards  team_b_red_cards  \
6287                   -1                -1                -1   
6363                   -1                -1                -1   
6376                   -1                -1                -1   
6296                   -1                -1                -1   
6319                   -1                -1                -1   
6455                   -1                -1                -1   
6440                   -1                -1                -1   
6372                   -1                -1                -1   
6306                   -1                -1                -1   
6337                   -1                -1                -1   

      team_a_shotsOnTarget  team_b_shotsOnTarget  team_a_shotsOffTarget  \
6287                    -1

In [10]:
# Check seasons from which matches are incomplete
data[data['status'] == 'incomplete' ]['season'].unique()

array(['2023/2024'], dtype=object)

In [11]:
# Check list of incomplete matches
data[data['status'] == 'suspended' ]['season'].unique()

array(['2023/2024'], dtype=object)

The above inspection confirms that all matches with the `incomplete` or `suspended` status are the current season. 

#### game_week and revised_game_week columns
The `revised_game_week` column indicates indicates whether the game-week (the week of the season when the match was played) was revised or not. We observe that for inspected sections of the dataframe, no matches appear to have revised match weeks.

In [12]:
# check for unique values in the 'revised_game_week' column
data['revised_game_week'].unique()

array([-1], dtype=int64)

 -1 is the only unique value in the `revised_game_week` column. Thus this column does not provide any useful information and will be dropped.
 As mentioned before, the `game_week`, indicates the week of the season when the match was played. Typically, the EPL games run from from game-week 1 to game-week 38 (assuming a team plays a single EPL game every week - which is usually the case). 

In [13]:
# check for unique values in the 'game_week' column
data['game_week'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38,  0], dtype=int64)

The above output mostly confirms our that there are typically 38 game weeks in a season. However, there appears to be some entries where the column value is 0.

In [14]:
# check for entries where game_week == 0
data[data['game_week'] == 0]

id  homeID  awayID     season    status  roundID  game_week  \
4180  753648      59     142  2008/2009  complete    53612          0   
4181  753649     153     221  2008/2009  complete    53612          0   
4182  753650     147      92  2008/2009  complete    53612          0   
4183  753651     150     162  2008/2009  complete    53612          0   
4184  753652     144     216  2008/2009  complete    53612          0   
...      ...     ...     ...        ...       ...      ...        ...   
4935  755509     153     158  2007/2008  complete    53621          0   
4936  755510      92     151  2007/2008  complete    53621          0   
4937  755511     156      59  2007/2008  complete    53621          0   
4938  755512     206     216  2007/2008  complete    53621          0   
4939  755513     144     157  2007/2008  complete    53621          0   

      revised_game_week                     homeGoals           awayGoals  \
4180                 -1                         ['4']                  []   
4181                 -1                   ['4', '10']              ['48']   
4182                 -1                  ['71', '87']              ['90']   
4183                 -1                  ['23', '81']               ['8']   
4184                 -1                  ['44', '64']  ['21', '66', '90']   
...                 ...                           ...                 ...   
4935                 -1                   ['8', '88']        ['14', '58']   
4936                 -1                            []        ['69', '74']   
4937                 -1                            []              ['24']   
4938                 -1  ['32', '73', '90+1', '90+3']              ['49']   
4939                 -1            ['28', '70', '82']              ['46']   

      homeGoalCount  awayGoalCount  totalGoalCount  team_a_corners  \
4180              1              0               1              -1   
4181              2              1               3              -1   
4182              2              1               3              -1   
4183              2              1               3              -1   
4184              2              3               5              -1   
...             ...            ...             ...             ...   
4935              2              2               4              -1   
4936              0              2               2              -1   
4937              0              1               1              -1   
4938              4              1               5              -1   
4939              3              1               4              -1   

      team_b_corners  totalCornerCount  team_a_offsides  team_b_offsides  \
4180              -1                -1               -1               -1   
4181              -1                -1               -1               -1   
4182              -1                -1               -1               -1   
4183              -1                -1               -1               -1   
4184              -1                -1               -1               -1   
...              ...               ...              ...              ...   
4935              -1                -1               -1               -1   
4936              -1                -1               -1               -1   
4937              -1                -1               -1               -1   
4938              -1                -1               -1               -1   
4939              -1                -1               -1               -1   

      team_a_yellow_cards  team_b_yellow_cards  team_a_red_cards  \
4180                   -1                   -1                -1   
4181                    2                    1                 0   
4182                    1                    2                 0   
4183                    3                    0                 0   
4184                    2                    2                 0   
...                   ...                  ...

In [15]:
# check seasons where game_week == 0 occur
data[data['game_week'] == 0]['season'].unique()

array(['2008/2009', '2007/2008'], dtype=object)

In [16]:
# check season unique game_week values for the 2008/2009 season
data[data['season'] == '2008/2009']['game_week'].unique()

array([0], dtype=int64)

In [17]:
# check season unique game_week values for the 2007/2008 season
data[data['season'] == '2007/2008']['game_week'].unique()

array([0], dtype=int64)

In [18]:
# check season unique game_week values for the 2009/2010 season
data[data['season'] == '2009/2010']['game_week'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 17, 14, 15, 16,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38], dtype=int64)

Further inspection shows that matches from the `2007/2008` and `2008/2009` seasons have all values in the `game_week` column as 0. `game_week` information seems to have been stardardized starting from `2009/2010` season. If `game_week` data from `2007/2008` and `2008/2009` seasons is to be used during EDA or modelling, it will be important to indicate the correct match-week information.

#### homeGoals and awayGoals columns
This columns consist of arrays that store goal timing data (when goals were scored during the match). Upon closer scrutiny of the dataset, it was observed that the columns `homeGoals` and `homeGoals_timings` contained identical information, as did `awayGoals` and `awayGoals_timings`. In the interest of eliminating redundancy and retaining more descriptive columns, it makes sense to remove the `homeGoals` and `awayGoals` columns.

In [19]:
data[['homeGoals', 'awayGoals', 'homeGoals_timings', 'awayGoals_timings' ]].head()

homeGoals awayGoals homeGoals_timings awayGoals_timings
0  ['45+1', '57']    ['47']    ['45+1', '57']            ['47']
1              []    ['82']                []            ['82']
2              []    ['74']                []            ['74']
3           ['5']    ['59']             ['5']            ['59']
4          ['11']    ['67']            ['11']            ['67']

#### corner & offside columns
Dataframe snapshots (head, tail, sample) showed that some seasons, particularly seasons priror to the `2013/2014` season had missing data (encoded as -1) in the following columns 
- `team_a_corners`
- `team_b_corners`
- `totalCornerCount`
- `team_a_offsides`
- `team_b_offsides`
- `team_a_shotsOnTarget`
- `team_b_shotsOnTarget`
- `team_a_shotsOffTarget`
- `team_b_shotsOffTarget`
- `team_a_shots`
- `team_b_shots`
- `team_a_fouls`
- `team_b_fouls`
- `team_a_possession`
- `team_b_possession`

In [20]:
# group by season and get descriptive stats for columns suspected to contain missing data
missing_data_columns = [
                        'team_a_corners',
                        'team_b_corners',
                        'totalCornerCount',
                        'team_a_offsides',
                        'team_b_offsides',
                        'team_a_shotsOnTarget',
                        'team_b_shotsOnTarget',
                        'team_a_shotsOffTarget',
                        'team_b_shotsOffTarget',
                        'team_a_shots',
                        'team_b_shots',
                        'team_a_fouls',
                        'team_b_fouls',
                        'team_a_possession',
                        'team_b_possession'
                        ]

data.groupby('season')[missing_data_columns].describe()

team_a_corners                                                \
                   count      mean       std  min  25%  50%  75%   max   
season                                                                   
2007/2008          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2008/2009          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2009/2010          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2010/2011          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2011/2012          380.0 -1.000000  0.000000 -1.0 -1.0 -1.0 -1.0  -1.0   
2012/2013          380.0  6.252632  3.158833  0.0  4.0  6.0  8.0  15.0   
2013/2014          380.0  6.092105  2.962286  0.0  4.0  6.0  8.0  14.0   
2014/2015          380.0  6.042105  3.243555  0.0  4.0  6.0  8.0  18.0   
2015/2016          380.0  5.931579  3.247983  0.0  4.0  5.0  8.0  17.0   
2016/2017          380.0  5.673684  3.085217  0.0  3.0  5.0  8.0  19.0   
2017/2018          380.0  5.750000  2.999230  0.0  4.0  5.0  8.0  18.0   
2018/2019          380.0  5.726316  3.025632  0.0  4.0  5.0  8.0  17.0   
2019/2020          380.0  5.771053  3.212394  0.0  3.0  5.0  8.0  17.0   
2020/2021          380.0  5.536842  3.026917  0.0  3.0  5.0  7.0  16.0   
2021/2022          380.0  5.584211  3.014611  0.0  3.0  5.0  7.0  15.0   
2022/2023          380.0  5.610526  3.071856  0.0  3.0  5.0  8.0  17.0   
2023/2024          380.0  2.647368  4.138993 -1.0 -1.0  1.0  6.0  17.0   

          team_b_corners                                                 \
                   count      mean       std  min   25%  50%  75%   max   
season                                                                    
2007/2008          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2008/2009          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2009/2010          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2010/2011          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2011/2012          380.0 -1.000000  0.000000 -1.0 -1.00 -1.0 -1.0  -1.0   
2012/2013          380.0  4.792105  2.784511  0.0  3.00  4.0  6.0  17.0   
2013/2014          380.0  4.665789  2.695915  0.0  3.00  4.0  6.0  14.0   
2014/2015          380.0  4.663158  2.568782  0.0  3.00  4.0  6.0  13.0   
2015/2016          380.0  4.876316  2.509710  0.0  3.00  5.0  7.0  14.0   
2016/2017          380.0  4.742105  2.705767  0.0  3.00  4.0  6.0  15.0   
2017/2018          380.0  4.544737  2.769148  0.0  2.75  4.0  6.0  13.0   
2018/2019          380.0  4.557895  2.731487  0.0  2.75  4.0  6.0  14.0   
2019/2020          380.0  4.928947  2.839173  0.0  3.00  5.0  7.0  16.0   
2020/2021          380.0  4.613158  2.669151  0.0  3.00  4.0  6.0  13.0   
2021/2022          380.0  4.818421  2.772115  0.0  3.00  5.0  7.0  14.0   
2022/2023          380.0  4.455263  2.814512  0.0  2.00  4.0  6.0  19.0   
2023/2024          380.0  1.873684  3.397165 -1.0 -1.00  1.0  4.0  13.0   

          totalCornerCount                                                    \
                     count       mean       std  min  25%   50%    75%   max   
season                                                                         
2007/2008            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2008/2009            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2009/2010            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2010/2011            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2011/2012            380.0  -1.000000  0.000000 -1.0 -1.0  -1.0  -1.00  -1.0   
2012/2013            380.0  11.044737  3.681315  0.0  9.0  11.0  13.25  23.0   
2013/2014            380.0  10.757895  3.522909  2.0  8.0  11.0  13.00  21.0   
2014/2015            380.0  10.705263  3.521135  1.0  8.0  11.0  13.00  23.0   
2015/2016            380.0  10.807895  3.604811  1.0  8.0  10.0  13.00  25.0   
2016/2017            380.0  10.415789  3.442826  2.0  8.0  10.0  13.

The output above confirms that seasons prior to the 2013/2014 season indeed are missing data in the columns identified above. Given that shots, shots on target, and corner kicks are occasions in a football/soccer match when the probility of scoring is high, and therefore probably have significant match outcome prediction power, we will likely retain only data from 2013/2014 season onwards.

#### Referee ID column
The 'refereeID column is a unique identifier for the referee who officiated a match. Dataframe snapshots revealed that the referee column had some missing values:

In [21]:
data[data['refereeID'].isnull()].shape

(2784, 215)

In [22]:
data[data['refereeID'].isnull()]['season'].unique()

array(['2016/2017', '2015/2016', '2014/2015', '2013/2014', '2017/2018',
       '2012/2013', '2008/2009', '2007/2008', '2023/2024'], dtype=object)

Output indicates that 2,784 entries in the refereeID column have null values. Given that Referees are randomly selected and are often assumed to have very little influence on the match outcome (assuming match officiation is fair), we can safely drop the referee column.

#### Stadium Location
Examination of dataframe shows that the stadium location column has some missing values. since this column has limited relevance, we will add to the list of columns to be dropped.

In [23]:
data[data['stadium_location'].isnull()]['season'].unique()

array(['2016/2017', '2015/2016', '2014/2015', '2013/2014', '2018/2019',
       '2019/2020', '2011/2012', '2010/2011', '2009/2010', '2008/2009',
       '2007/2008', '2020/2021', '2021/2022', '2022/2023', '2023/2024'],
      dtype=object)

#### odds columns
The dataframe snapshots (head, tail, sample) revealed that the values of most of the odds columns, with the exception of `odds_ft_1`, `odds_ft_x`, & `odds_ft_1`  columns, were zeros. Odds represent the probability of a given match outcome, therefore, are not expected to be zero.Given that this columns essentially are missing data, we will drop them from the dataset.

In [24]:
# get list of odds columns 
list_of_columns = list(data.columns)

# odds columns
odds_columns = [column for column in list_of_columns if column.startswith('odds')]


In [25]:
# compute the % zero values for each column and store the results in a dictionary
zero_counts = {}
for column in odds_columns:
    zero_counts[column] = (data[column] == 0).sum()/6460

In [26]:
zero_counts

{'odds_ft_1': 0.3238390092879257,
 'odds_ft_x': 0.3238390092879257,
 'odds_ft_2': 0.3238390092879257,
 'odds_ft_over05': 0.6174922600619195,
 'odds_ft_over15': 0.6174922600619195,
 'odds_ft_over25': 0.6164086687306501,
 'odds_ft_over35': 0.6174922600619195,
 'odds_ft_over45': 0.6174922600619195,
 'odds_ft_under05': 0.6174922600619195,
 'odds_ft_under15': 0.6174922600619195,
 'odds_ft_under25': 0.6167182662538699,
 'odds_ft_under35': 0.6173374613003096,
 'odds_ft_under45': 0.6174922600619195,
 'odds_btts_yes': 0.6174922600619195,
 'odds_btts_no': 0.6174922600619195,
 'odds_team_a_cs_yes': 0.6188854489164086,
 'odds_team_a_cs_no': 0.6188854489164086,
 'odds_team_b_cs_yes': 0.6339009287925697,
 'odds_team_b_cs_no': 0.6335913312693499,
 'odds_doublechance_1x': 0.6195046439628483,
 'odds_doublechance_12': 0.6195046439628483,
 'odds_doublechance_x2': 0.6195046439628483,
 'odds_1st_half_result_1': 0.6185758513931888,
 'odds_1st_half_result_x': 0.6187306501547988,
 'odds_1st_half_result_2': 0.

The output above reveals that 32% of entries in the `odds_ft_1`, `odds_ft_x`, & `odds_ft_1`  columns have invalid odd data (zeros) while the other odd columns consists of between 62% and 100%  invalid data.Therefore, to ensure model integrity all odds columns except `odds_ft_1`, `odds_ft_x`, & `odds_ft_1`  will be dropped from the dataset.Entries with invalid data in the `odds_ft_1`, `odds_ft_x`, & `odds_ft_1`  columns will be dropped. 

#### winning_team column
Inspection revealed that this column indicates winning team using the teamid and draws using -1. For EDA it might be useful to create column that indicates home win with `1`, draws with `x` and away wins with 2, which is the industry standard.

#### no_home_away column
We observed tha the `no_home_away` appears to consist of mostly 0 which makes sense as this would only be 1 if a match was played on a neutral ground. This rarely happens in the EPL.

In [27]:
data['no_home_away'].unique()

array([0], dtype=int64)

The above output confirms that no games were played on neutral ground and thus this column can be dropped.
#### Attendance column
In this column we noted that zero attendance or missing attendance was encoded with -1. 

In [28]:
data[data['attendance']==-1]['attendance'].count()

1822

In [29]:
# List of seasons to filter out
season_not_relevant = ['2007/2008', '2008/2009', '2009/2010', '2010/2011', 
                       '2011/2012', '2012/2013', '2023/2024']

# Filter out rows where the "season" column is in the list
filtered_data = data[~data['season'].isin(season_not_relevant)]

# check missing attendance for relevant seasons
filtered_data[filtered_data['attendance']==-1]['attendance'].count()

729

In [30]:
filtered_data[filtered_data['attendance']==-1].groupby('season')['attendance'].count()

season
2013/2014     16
2014/2015     16
2015/2016     16
2016/2017     16
2020/2021     81
2021/2022    263
2022/2023    321
Name: attendance, dtype: int64

Further inspection reveals that for the relevant seasons, seasons to be considered to have sufficient data, very few matches between `2013` and `2019` had missing fan attendance information. While missing attendance data was expected for the `2020/2021` due covid restrictions, missing data in `2021/2022` and `2022/2023` is unusual. 
Since attendance data may be useful in predicting match outcome, missing data could be obtained from past attendance data or stadium capacity can be used with the assumption that the stadium was full.

#### goalTimingDisabled, corner_timings_recorded, & card_timings_recorded columns
This columns consisted mostly of 1's or 0's. API documentation revealed that this columns were used to indicate whether goal timing, corner timing or card timing data was captured. This information has little relevance and can be safely dropped dropped.

#### team_a_fh_corners	team_b_fh_corners	team_a_2h_corners	team_b_2h_corners	corner_fh_count	corner_2h_count	team_a_fh_cards	team_b_fh_cards	team_a_2h_cards	team_b_2h_cards	total_fh_cards	total_2h_cards 

- Review this columns (2016/2017) season showed some missing data for these this season

#### attacks_recorded column
The attacks_recorded column tracks whether or not attack data was captured encoded as -1 (attack data not recorded) and 1 (attack data was recorded). This is a metadata column and thus can be dropped as it does have any prective relevance. 

In [31]:
filtered_data['attacks_recorded'].unique()

array([-1,  1], dtype=int64)

In [32]:
filtered_data['attacks_recorded'].value_counts(normalize=True)

 1    0.690263
-1    0.309737
Name: attacks_recorded, dtype: float64

The output above that approximately 56% of matches are missing attack data. The affected columns include:
- `team_a_dangerous_attacks`	
- `team_b_dangerous_attacks`	
- `team_a_attacks`	
- `team_b_attacks`
Given that a large fraction of the attack data is missing, and because this data cannot be imputed, dropping the affected columns is a sensible approach. Additionally, given that live match predictions, which rely on attack data,  is not part of the project scope, dropping attack data does not significantly affect the success of the project. 

In [33]:
data['competition_id'].unique()

array([   9,   10,   11,   12,  161,  246, 1625, 2012, 3119, 3121, 3125,
       3131, 3137, 4759, 6135, 7704, 9660], dtype=int64)